src/swupd/hash_test.go

```diff
@@ -87,3 +89,78 @@ func TestHashEqual(t *testing.T) {
 
 // Tip, to generate random hash values use this.
 // hexdump -n32 -e '32 "%02x" "\n"' /dev/random
+
+const (
+	Dir  = 0040000
+	Reg  = 0100000
+	Link = 0120000
+)
+
+func TestGenHash(t *testing.T) {
+	testCases := []struct {
+		info   syscall.Stat_t
+		data   []byte
+		result string
+	}{
+		{syscall.Stat_t{Mode: (Dir + 0755)},
+			[]byte("DIRECTORY"), directoyhash},
+		{syscall.Stat_t{Mode: (Dir + 01777)},
+			[]byte("DIRECTORY"), "d93a5e9129361e28b9e244fe422234e3a1794b001a082aeb78e16fd881673a2b"},
+		{syscall.Stat_t{Mode: Reg + 0644, Uid: 1000, Gid: 1000},
+			[]byte(""), "b85f1dc2c2317a20f47a36d3257313b131124ffa6d4f19bb060d43014fd386b0"},
+	}
+
+	for _, tc := range testCases {
+		r := genHash(tc.info, tc.data)
+		if r != tc.result {
+			t.Errorf("Unexpected result %s for %v", r, tc)
+		}
+	}
+}
+
+const (
+	// hash for a rwxr-xr-x root owned directory
+	directoyhash = "6c27df6efcd6fc401ff1bc67c970b83eef115f6473db4fb9d57e5de317eba96e"
+	missinghash  = "0000000000000000000000000000000000000000000000000000000000000000"
+)
+
+func TestHashcalc(t *testing.T) {
+	testCases := []struct {
+		filename string
+		result   string
+	}{
+		{"/", directoyhash},
+		{"/does not exist", missinghash},
+		{"/usr", directoyhash},
+		{"/usr/share/doc/systemd/LICENSE.GPL2", "d9d34a1e44f3684286dd07c6a9e1747a1307e4421cd5d70f71a548c446a9ca54"},
```

Comment:
So my inclination here is to use files located from the source directory specificly for these tests instead of relying on files on the host. Is there a problem with that I'm overlooking?

Reply:
I think it is reasonable to want to check a known file on the filesystem with known contents to make sure we aren't testing inside an echo chamber. This way we can make sure swupd-client calculates the same hash as well.

Reply:
I get annoyed with packages that can't run tests in a container that is specificly existing to build go code and these kinds of tests are exactly that.

Reply:
sure, maybe this test doesn't need to be part of the unit tests, but this test should exist somewhere. Maybe our internally-carried swupd integration tests?

Reply:
@bryteise Yes, you are overlooking that I don't know what user id I am running as. The hash encodes various properties including the UID and GID of the file.

Every since BSD added file quotas, the 'chown' system call became privileged and so I can't  change the file to a known user, so I have to have an existing known file, no file I can create will allow this.

The alternative is to have the unit tests call out to sudo. Personally I am not happy with this idea.

Reply:
Hrm well the tests executed in travis wouldn't have a problem with doing that but I don't really like needing that either. Perhaps requiring the tests to have unshare available (or be run under sudo) could be a possibility but =P.

Reply:
Perhaps run under fakeroot or pseudo (https://www.yoctoproject.org/tools-resources/projects/pseudo)? fakeroot should be good enough and is readily available in distros.

Reply:
@pohly Thank you for the suggestion. However fakeroot (and pseudo according to its documentation although I haven't checked it) are not going to be good enough. Fakeroot (and apparently pseudo) work by injecting a library using LD_PRELOAD, which provides wrappers around various system calls made via functions in the glibc library. The toolchain we are using produces static binaries, so the dynamic linker is not called, so nothing honors the LD_PRELOAD environment variable, and even if it did we are not using glibc but making the system calls directly.

It would be possible to do this differently using the ptrace debugger interface. Tell the os to trap on entry to the system calls we are interested in, and alter the parameters, then trap on return and alter the return values accordingly. This is quite a lot more work than fakeroot & pseudo need to do. I have just done a web search and found https://fakeroot-ng.lingnu.com/index.php/Home_Page which claims to do this.

@bryteise In order to run unshare(1) successfully you need to be root, unless you know something that I don't (entirely possible).

Reply:
unshare --map-root-user --user sh -c whoami should get root as the output when run as a non-root user. I expect the only thing needed is to get a consistent uid/gid reported which that would do.